x-pack/plugins/infra/public/observability_logs/log_stream_query_state/src/types.ts

```diff
@@ -36,38 +37,78 @@ export interface LogStreamQueryContextWithValidationError {
   validationError: Error;
 }
 
+export type ExtendedTimeRange = TimeRange & { lastChangedCompletely: number };
+export interface LogStreamQueryContextWithTimeRange {
+  timeRange: ExtendedTimeRange;
+}
+
+export interface LogStreamQueryContextWithRefreshInterval {
+  refreshInterval: RefreshInterval;
+}
+
+export interface Timestamps {
+  startTimestamp: number;
+  endTimestamp: number;
+  lastChangedTimestamp: number;
+}
+
+export interface LogStreamQueryContextWithTimestamps {
+  timestamps: Timestamps;
+}
+
+export type LogStreamQueryContextWithTime = LogStreamQueryContextWithTimeRange &
+  LogStreamQueryContextWithRefreshInterval &
+  LogStreamQueryContextWithTimestamps;
+
 export type LogStreamQueryTypestate =
   | {
       value: 'uninitialized';
-      context: LogStreamQueryContextWithDataViews;
+      context: LogStreamQueryContextWithDataViews & LogStreamQueryContextWithTime;
     }
   | {
-      value: 'hasQuery' | { hasQuery: 'validating' };
+      value: 'query' | { query: 'validating' };
       context: LogStreamQueryContextWithDataViews &
         LogStreamQueryContextWithParsedQuery &
         LogStreamQueryContextWithQuery &
-        LogStreamQueryContextWithFilters;
+        LogStreamQueryContextWithFilters &
+        LogStreamQueryContextWithTime;
     }
   | {
-      value: { hasQuery: 'valid' };
+      value: { query: 'valid' };
       context: LogStreamQueryContextWithDataViews &
         LogStreamQueryContextWithParsedQuery &
         LogStreamQueryContextWithQuery &
-        LogStreamQueryContextWithFilters;
+        LogStreamQueryContextWithFilters &
+        LogStreamQueryContextWithTime;
     }
   | {
-      value: { hasQuery: 'invalid' };
+      value: { query: 'invalid' };
       context: LogStreamQueryContextWithDataViews &
         LogStreamQueryContextWithParsedQuery &
         LogStreamQueryContextWithQuery &
         LogStreamQueryContextWithFilters &
+        LogStreamQueryContextWithTime &
         LogStreamQueryContextWithValidationError;
+    }
+  | {
+      value: 'time';
```

Comment:
For correct typing of `.matches()` checks we'd need to supply all nested states here as well, I think:

```suggestion
      value: 'time' | { time: 'initialized' } | { time: 'streaming' } | { time: 'static' };
```